### PR TITLE
Fix default deny port removal issue

### LINKED ISSUE #1782: new networkpolicy validation results: Egress policys freeze
Hi ovn k8s ! looks like we have some failing network policy tests in the new framework

See https://github.com/kubernetes/kubernetes/pull/91592 for methodology on reproducing these
 
These tests pass on calico, antrea, and cillium alike, so im assuming they are bugs in OVN-k8s. 

Would like to get some of these evaluated

**UPDATE (11/5)** 

- There was one important failure that was a core issue, now being fixed
```
 should stop enforcing policies after they are deleted
```


- most of these were related to named ports..
```
                                                                                                                                                                                                                                                                                                                                                                                                                    
Summarizing 7 Failures:                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                               
[Fail] [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client [It] should allow egress access on one named port [Feature:Netpol]                                                                                                                                             
/home/ubuntu/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:487                                                                                                                                                                            
                                                                                                                                                                                                                                                                                               
[Fail] [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client [It] should work with Ingress, Egress specified together [Feature:Netpol]                                                                                                                                      
/home/ubuntu/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:617                                                                                                                                                                            
                                                                                                                                                                                                                                                                                               
[Fail] [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client [It] should stop enforcing policies after they are deleted [Feature:Netpol]                                                                                                                                    
/home/ubuntu/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:701                                                                                                                                                                            
                                                                                                                                                                                                                                                                                               
[Fail] [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client [It] should allow ingress access on one named port [Feature:Netpol]                                                                                                                                                                                                                                                                                    
/home/ubuntu/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:448                                                                                                                                                                            
                                                                                                                                                                                                                                                                                               
[Fail] [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client [It] should enforce multiple ingress policies with ingress allow-all policy taking precedence [Feature:Netpol]                                                                                                 
/home/ubuntu/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:650                                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                                                                                       
[Fail] [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client [It] should allow ingress access from namespace on one named port [Feature:Netpol]                                                                                                                                                                                                                                                                     
/home/ubuntu/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:472                                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                               
[Fail] [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client [It] should enforce multiple egress policies with egress allow-all policy taking precedence [Feature:Netpol]                                                                                                                                                                                                                                           
/home/ubuntu/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:670                                                                                  

```

### PR DESCRIPTION
**Pull request by astoycos**
Fixes #1782

With the code updates to fix problems with orphaned ACLs 
in PR #1665 the code for `localPodDelDefaultDeny` was 
changed in order to be able to clean up related 
resources from both `policy.go` and `namespace.go` 

In order to do so, the `knet.policy` object could no longer
be used in the function call since we didn't have access
to it in `namespace.go` 

Therefore the logic used to determined `PolicyType` was changed
to use only the existing `namespacePolicy` fields egressPolicies and 
ingresPolicies. However it had some issues and the number of egress/ingress
rules weren't enough to tell the `NetworkPolicyType` 

For example, if we had a `NetworkPolicy`
that blocked ALL traffic by default but did not contain any 
ingress/egress rules, the new logic would fail to delete the ports from the 
default deny port group 

FIX: 
Add a `policyType:      policy.Spec.PolicyTypes,`  ensure we can always determine 
the type of `NetworkPolicy` regardless of the number of ingress and egress rules

### CODE REVIEW

File: go-controller/pkg/ovn/policy.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ type namespacePolicy struct {
 	sync.Mutex
 	name            string
 	namespace       string
+	policyType      []knet.PolicyType
```
Comment (trozet) [inline go-controller/pkg/ovn/policy.go:22]: should be "policyTypes"

File: go-controller/pkg/ovn/policy.go [hunk 2]
```diff
@@ -419,7 +421,9 @@ func (oc *Controller) localPodDelDefaultDeny(
 	oc.lspMutex.Lock()
 	defer oc.lspMutex.Unlock()
 
-	if !(len(np.ingressPolicies) == 0 && len(np.egressPolicies) > 0) {
+	//Remove port from ingress deny port-group for [Ingress] and [ingress,egress] PolicyTypes
```
Comment (trozet) [inline go-controller/pkg/ovn/policy.go:424]: nit: add a space after starting a // comment

Comment (trozet) [reply go-controller/pkg/ovn/policy.go:424]: btw thanks for adding the comments

**Review by trozet — CHANGES_REQUESTED** [review]
minor syntax changes, otherwise looks good

**Comment by astoycos** [comment]
@trozet All set!

**Comment by trozet** [comment]
interesting @astoycos @dcbw looks like there is a race when we delete address set, and then another thread tries to add a peer address to a now nil address set object:

```
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*gressPolicy).destroy()
      /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/gress_policy.go:422 +0xac

  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*gressPolicy).addPeerPod()
      /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/gress_policy.go:104 +0xba
```

Looks like we need to check if the gp.peerAddressSet is nil before try to add addresses to it. I think I recently added a check like that in ensurePeerAddressSet. @astoycos can you add another commit here and do a similar check in addPeerPod?

**Comment by astoycos** [comment]
@trozet sounds good will do

File: go-controller/pkg/ovn/gress_policy.go [hunk 3]
```diff
@@ -96,6 +96,10 @@ func (gp *gressPolicy) ensurePeerAddressSet(factory AddressSetFactory) error {
 }
 
 func (gp *gressPolicy) addPeerPod(pod *v1.Pod) error {
+	if gp.peerAddressSet == nil {
+		return fmt.Errorf("peer AddressSet is nil, cannot add peer pod: %s", pod.ObjectMeta.Name)
```
Comment (trozet) [inline go-controller/pkg/ovn/gress_policy.go:100]: maybe add the policy name here as well?

**Review by trozet — CHANGES_REQUESTED** [review]

**Review by trozet — APPROVED** [review]

**Pull request merged**